File: 6 linked list/DoublyLinkedList.js
// Constructor function that creates a doubly linked list
function DoublyLinkedList() {
  this.head = null;
  this.tail = null;
}

// Constructor function that creates a new doubly linked node
function DoublyNode(val) {
  this.value = val;
  this.prev = null;
  this.next = null;
}

// Adds node to the end of the doubly linked list
DoublyLinkedList.prototype.push = function (value) {
  // Create new DoublyNode
  const newNode = new DoublyNode(value);
  // Check if head exists
  if (!this.head) {
    this.head = newNode;
    this.tail = newNode;
  } else {
    // Point the tail's next to the new node
    this.tail.next = newNode;
    // Point the new node's prev to the current tail
    newNode.prev = this.tail;
    // Set the new node as the new tail
    this.tail = newNode;
  }
};

// Contains function remains the same

// Bonus
// Adds node to the beginning of the doubly linked list
DoublyLinkedList.prototype.addToHead = function (value) {
  const newNode = new DoublyNode(value);
  if (!this.head) {
    this.head = newNode;
    this.tail = newNode;
  } else {
    newNode.next = this.head;
    this.head.prev = newNode;
    this.head = newNode;
  }
};

// Extra Bonus
// Insert an item at the position specified in the doubly linked list
DoublyLinkedList.prototype.insert = function (value, position) {
  if (position < 0) {
    throw new Error("Position cannot be negative.");
  }

  if (position === 0) {
    this.addToHead(value);
    return;
  }

  const newNode = new DoublyNode(value);
  let current = this.head;
  let currentPosition = 0;

  while (current !== null && currentPosition < position) {
    current = current.next;
    currentPosition++;
  }

  if (currentPosition < position) {
    throw new Error("Position is beyond the length of the list.");
  }

  newNode.prev = current.prev;
  newNode.next = current;

  if (current.prev) {
    current.prev.next = newNode;
  } else {
    this.head = newNode;
  }

  current.prev = newNode;

  if (current === null) {
    this.tail = newNode;
  }
};

// Extra Bonus
// Remove the first occurrence of value from the doubly linked list
DoublyLinkedList.prototype.removeItem = function (value) {
  if (!this.head) {
    return;
  }

  let current = this.head;

  while (current !== null) {
    if (current.value === value) {
      if (current.prev) {
        current.prev.next = current.next;
      } else {
        this.head = current.next;
      }

      if (current.next) {
        current.next.prev = current.prev;
      } else {
        this.tail = current.prev;
      }

      return;
    }

    current = current.next;
  }
};

// Extra Bonus
// Remove element at the specified position in the doubly linked list
DoublyLinkedList.prototype.removePosition = function (position) {
  if (!this.head || position < 0) {
    return;
  }

  if (position === 0) {
    this.head = this.head.next;
    if (this.head === null) {
      this.tail = null;
    } else {
      this.head.prev = null;
    }
    return;
  }

  let current = this.head;
  let currentPosition = 0;

  while (current !== null && currentPosition < position) {
    current = current.next;
    currentPosition++;
  }

  if (currentPosition < position || current === null) {
    return; // Position is beyond the length of the list.
  }

  if (current.prev) {
    current.prev.next = current.next;
  } else {
    this.head = current.next;
  }

  if (current.next) {
    current.next.prev = current.prev;
  } else {
    this.tail = current.prev;
  }
};

// Bonus
// Reverse the doubly linked list
DoublyLinkedList.prototype.reverse = function () {
  let current = this.head;
  while (current !== null) {
    // Swap next and prev pointers for current node
    const temp = current.next;
    current.next = current.prev;
    current.prev = temp;

    // Move to the next node in the original list
    current = temp;
  }

  // Swap head and tail pointers to reflect the reversed list
  const tempHead = this.head;
  this.head = this.tail;
  this.tail = tempHead;
};

File: 6 linked list/doubly-linked-list.py
class Node:
   def __init__(self, data):
      self.data = data
      self.next = None
      self.prev = None
        
class doubly_linked_list:
   def __init__(self):
      self.head = None
      
# Define the append method to add elements at the end
   def append(self, NewVal):
      NewNode = Node(NewVal)
      NewNode.next = None
      if self.head is None:
         NewNode.prev = None
         self.head = NewNode
         return
     #move to end if no self.tail in LL constructor
      last = self.head
      while (last.next is not None):
         last = last.next
      last.next = NewNode
      NewNode.prev = last
      return
  
# Adding data elements at head		
   def push(self, NewVal):
      NewNode = Node(NewVal)
      NewNode.next = self.head
      if self.head is not None:
         self.head.prev = NewNode
      self.head = NewNode

# Define the insert method to insert the element		
   def insert(self, prev_node, NewVal):
      if prev_node is None:
         return
      NewNode = Node(NewVal)
      NewNode.next = prev_node.next
      prev_node.next = NewNode
      NewNode.prev = prev_node
      if NewNode.next is not None:
         NewNode.next.prev = NewNode

# Print the Doubly Linked list		
   def listprint(self, node):
      while (node is not None):
         print(node.data),
         last = node
         node = node.next

dllist = doubly_linked_list()
dllist.push(12)
dllist.push(8)
dllist.push(62)
dllist.insert(dllist.head.next, 13)
dllist.listprint(dllist.head)